pkg/networkservice/common/refresh/client.go

```diff
@@ -55,76 +56,72 @@ func (t *refreshClient) Request(ctx context.Context, request *networkservice.Net
 	if err != nil {
 		return nil, err
 	}
-	// Clone the request
-	req := request.Clone()
-	// Set its connection to the returned connection we received
-	req.Connection = rv
-
-	expire, err := t.getExpireDuration(request)
+	expireTime, err := ptypes.Timestamp(request.GetConnection().GetPath().GetPathSegments()[request.GetConnection().GetPath().GetIndex()].GetExpires())
 	if err != nil {
-		return nil, errors.Wrapf(err, "Error creating timer from Request.Connection.Path.PathSegment[%d].ExpireTime", request.GetConnection().GetPath().GetIndex())
+		return nil, err
 	}
+
+	// Create refreshRequest
+	refreshRequest := request.Clone()
+	refreshRequest.GetConnection().Context = rv.Context
+	refreshRequest.GetConnection().Mechanism = rv.Mechanism
+	refreshRequest.GetConnection().NetworkServiceEndpointName = rv.NetworkServiceEndpointName
+
+	// TODO - introduce random noise into duration avoid timer lock
+	duration := time.Until(expireTime) / 3
 	t.executor.AsyncExec(func() {
-		id := req.GetConnection().GetId()
-		// check if it is refresh request
-		if refreshCtx := refreshContext(ctx); refreshCtx != nil {
-			// refresh was canceled
-			if refreshCtx.Err() != nil {
-				return
-			}
-			// we reuse non-canceled refresh context for the next refresh request
-			timer := t.createTimer(ctx, req, expire, opts...)
-			t.connectionTimers[id] = timer
-		} else {
-			// cancel refresh of previous request if any
-			if timer, ok := t.connectionTimers[id]; ok {
-				timer.Stop()
-			}
-			if canceller, ok := t.refreshCancellers[id]; ok {
-				canceller()
-			}
-			// add refresh context to request context
-			refreshCtx, cancelFunc := context.WithCancel(context.Background())
-			newCtx := withRefreshContext(ctx, refreshCtx)
+		// Create the refresh context
+		var cancel context.CancelFunc
+		refreshCtx, cancel := context.WithCancel(extend.WithValuesFromContext(t.ctx, ctx))
+		if deadline, ok := ctx.Deadline(); ok {
+			refreshCtx, cancel = context.WithDeadline(refreshCtx, deadline.Add(duration))
+		}
 
-			timer := t.createTimer(newCtx, req, expire, opts...)
-			t.connectionTimers[id] = timer
-			t.refreshCancellers[id] = cancelFunc
+		// Stop any existing timers
+		if timer, ok := t.timers[request.GetConnection().GetId()]; ok {
+			timer.Stop()
 		}
+
+		// Set new timer
+		var timer *time.Timer
+		timer = time.AfterFunc(duration, func() {
+			t.executor.AsyncExec(func() {
+				// Check to see if we've been superseded by another timer, if so, do nothing
+				currentTimer, ok := t.timers[refreshRequest.GetConnection().GetId()]
+				if ok && currentTimer != timer {
+					cancel()
+					return
+				}
+				// Resend request if the refreshCtx is not Done.  Note: refreshCtx should only be done if
+				// t.ctx has been canceled (usually because the clientCtx has been canceled
+				select {
+				case <-refreshCtx.Done():
+				default:
+					if _, err := t.Request(refreshCtx, refreshRequest, opts...); err != nil {
```

Comment:
Do we block all other requests by doing Request inside executor?

Reply:
Good catch.  Fixed!